pkg/telemetrylogs/json_condition_builder.go

```diff
@@ -0,0 +1,455 @@
+	// Build traversal + terminal recursively per-hop
+	compiled, err := c.recurseArrayHops(plan, operator, value, sb)
+	if err != nil {
+		return "", err
+	}
+
+	// sb.AddWhereClause(sqlbuilder.NewWhereClause().AddWhereExpr(sb.Args, compiled))
+	return compiled, nil
+}
+
+// buildTerminalCondition creates the innermost condition
+func (c *conditionBuilder) buildTerminalCondition(node *telemetrytypes.JSONAccessNode, operator qbtypes.FilterOperator, value any, sb *sqlbuilder.SelectBuilder) (string, error) {
+	// Use the parent's alias + current field name for the full path
+	fieldPath := node.FieldPath()
+
+	if node.TerminalConfig.ElemType.IsArray {
+		// switch operator for array membership checks
+		switch operator {
+		case qbtypes.FilterOperatorContains, qbtypes.FilterOperatorIn:
+			operator = qbtypes.FilterOperatorEqual
+		case qbtypes.FilterOperatorNotContains, qbtypes.FilterOperatorNotIn:
+			operator = qbtypes.FilterOperatorNotEqual
+		}
+		arrayCond, err := c.buildArrayMembershipCondition(node, operator, value, sb)
+		if err != nil {
+			return "", err
+		}
+		return arrayCond, nil
+	}
+	conditions := []string{}
+
+	elemType := node.TerminalConfig.ElemType
+	fieldExpr := fmt.Sprintf("dynamicElement(%s, '%s')", fieldPath, elemType.StringValue())
+	fieldExpr, value = querybuilder.DataTypeCollisionHandledFieldName(node.TerminalConfig.Key, value, fieldExpr, operator)
+
+	indexed := slices.ContainsFunc(node.TerminalConfig.Key.Indexes, func(index telemetrytypes.JSONDataTypeIndex) bool {
+		return index.Type == elemType && index.ColumnExpression == fieldPath
+	})
+	if elemType.IndexSupported && indexed {
+		indexedExpr := assumeNotNull(fieldPath, elemType)
+		emptyValue := func() any {
+			switch elemType {
+			case telemetrytypes.String:
+				return ""
+			case telemetrytypes.Int64, telemetrytypes.Float64, telemetrytypes.Bool:
+				return 0
+			default:
+				return nil
+			}
+		}()
+
+		// switch the operator and value for exists and not exists
+		switch operator {
+		case qbtypes.FilterOperatorExists:
+			operator = qbtypes.FilterOperatorNotEqual
+			value = emptyValue
+		case qbtypes.FilterOperatorNotExists:
+			operator = qbtypes.FilterOperatorEqual
+			value = emptyValue
+		default:
+			// do nothing
+		}
+
+		cond, err := c.applyOperator(sb, indexedExpr, operator, value)
+		if err != nil {
+			return "", err
+		}
+		conditions = append(conditions, cond)
+		// Switch operator to EXISTS
+		operator = qbtypes.FilterOperatorExists
+	}
+
+	cond, err := c.applyOperator(sb, fieldExpr, operator, value)
+	if err != nil {
+		return "", err
+	}
+	conditions = append(conditions, cond)
+	if len(conditions) > 1 {
+		return sb.And(conditions...), nil
+	}
+	return cond, nil
+}
+
+// buildArrayMembershipCondition handles array membership checks
+func (c *conditionBuilder) buildArrayMembershipCondition(node *telemetrytypes.JSONAccessNode, operator qbtypes.FilterOperator, value any, sb *sqlbuilder.SelectBuilder) (string, error) {
```

Comment:
note: Array membership check from client can come in form of `has(config.features, 'my-feature'), it should be handled.